### PR TITLE
Cy 1447 improve agents list

### PR DESCRIPTION
**Pull request by glukhman**

### CODE REVIEW

File: tests/integration_tests/tests/agent_tests/test_list_agents.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+########
+# Copyright (c) 2016 GigaSpaces Technologies Ltd. All rights reserved
```
Comment (inbala) [inline tests/integration_tests/tests/agent_tests/test_list_agents.py]: Change to `2013-2019 Cloudify `

File: rest-service/manager_rest/resource_manager.py [hunk 2]
```diff
@@ -1663,74 +1662,30 @@ def create_tasks_graph(self, name, execution_id, operations=None):
...
-        return True
+            filters['node_id'] = node_ids
+        if node_instance_ids is not None:
+            filters['node_instance_id'] = node_instance_ids
+        if install_method is not None:
+            filters['install_method'] = install_method
+        agents_raw = self.sm.list(models.Agent,
+                                  filters=filters,
+                                  all_tenants=all_tenants)
+        agents = [dict(id=agent.id,
```
Comment (tehasdf) [inline rest-service/manager_rest/resource_manager.py]: any reason to create a new list of dicts rather than just returning agents_raw directly?

File: tests/integration_tests/tests/agent_tests/test_list_agents.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        deployment1, _ = self.deploy_application(resource(
+            "dsl/agent_tests/with_agent.yaml"), deployment_id='d1')
+        deployment2, _ = self.deploy_application(resource(
+            "dsl/agent_tests/with_agent.yaml"), deployment_id='d2')
+
+        agent_list = self.client.agents.list()
+        self.assertEqual(agent_list.metadata['pagination']['total'], 2)
+        self.assertEqual(agent_list.items[0]['deployment'], 'd1')
+        self.assertEqual(agent_list.items[1]['deployment'], 'd2')
+        self.assertEqual(len(agent_list.items), 2)
```
Comment (tehasdf) [inline tests/integration_tests/tests/agent_tests/test_list_agents.py]: nit: in general assert on the length first, before touching `agent_list.items[0]`, so that if the resulting list is empty (or with just 1 item), we get a test failure rathen than an error

(also applies to other cases in this file)

File: tests/integration_tests/tests/agent_tests/test_list_agents.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        deployment2 = self.deploy(resource("dsl/agent_tests/with_agent.yaml"),
+                                  deployment_id='ns2')
+
+        agent_list = self.client.agents.list()
+        self.assertEqual(agent_list.metadata['pagination']['total'], 1)
+        self.assertEqual(len(agent_list.items), 1)
+        self.undeploy_application(deployment1.id)
+        self.delete_deployment(deployment2.id, validate=True)
+
+    def test_list_agents_all_tenants(self):
```
Comment (tehasdf) [inline tests/integration_tests/tests/agent_tests/test_list_agents.py]: multitenant test needs to be either in premium, or with a `@pytest.mark.skipif` checking for premium, because this will just error out on community

File: tests/integration_tests/tests/agent_tests/test_list_agents.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    def test_list_agents(self):
+        deployment1, _ = self.deploy_application(resource(
+            "dsl/agent_tests/with_agent.yaml"), deployment_id='d1')
+        deployment2, _ = self.deploy_application(resource(
+            "dsl/agent_tests/with_agent.yaml"), deployment_id='d2')
+
+        agent_list = self.client.agents.list()
+        self.assertEqual(agent_list.metadata['pagination']['total'], 2)
+        self.assertEqual(agent_list.items[0]['deployment'], 'd1')
+        self.assertEqual(agent_list.items[1]['deployment'], 'd2')
```
Comment (tehasdf) [inline tests/integration_tests/tests/agent_tests/test_list_agents.py]: i dont think the query has an order to it, so it's possible that postgres gives you `d2` first, then `d1` (most likely not in tests where there's just 2 items, but in theory possible)

better to `assertEqual({agent['deployment'] for agent in agent_list.items}, {'d1', 'd2'})`

(also applies to other cases in this file)

File: rest-service/manager_rest/rest/resources_v3_1/agents.py [hunk 6]
```diff
@@ -27,32 +26,29 @@
...
+    def get(self, _include=None, filters=None, pagination=None, sort=None,
+            all_tenants=None):
+        return get_storage_manager().list(
+            models.Agent,
+            include=_include,
+            filters=filters,
+            pagination=pagination,
+            sort=sort,
+            all_tenants=all_tenants,
+            get_all_results=True
```
Comment (inbala) [inline rest-service/manager_rest/rest/resources_v3_1/agents.py]: get_all_results should be a parameter received and default to false :
```
        get_all_results = verify_and_convert_bool(
            '_get_all_results',
            request.args.get('_get_all_results', False)
        )
```
It probably should have been a decorator but I don't think we should change it now.

File: tests/integration_tests/tests/agent_tests/test_list_agents.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+from integration_tests import AgentTestCase
+from integration_tests.tests.utils import (get_resource as resource,
+                                           create_rest_client)
+
+
+def _is_community():
```
Comment (inbala) [inline tests/integration_tests/tests/agent_tests/test_list_agents.py]: Maybe you can use instead of this function: `from manager_rest import premium_enabled`

File: rest-service/manager_rest/storage/resource_models.py [hunk 8]
```diff
@@ -657,6 +665,19 @@ def node_instance(cls):
...
     deployment_id = association_proxy('node_instance', 'deployment_id')
 
+    @property
+    def host_id(self):
+        return self.name
+
+    node = node_id
+    deployment = deployment_id
+    node_ids = node_id
+    node_instance_ids = node_instance_id
```
Comment (inbala) [inline rest-service/manager_rest/storage/resource_models.py]: 1. Make sure it doesn't affect the migration file.
2. Maybe the fields relevant to the response can be initialise  in the `to_response` function?

File: rest-service/manager_rest/rest/resources_v3_1/agents.py [hunk 9]
```diff
@@ -27,32 +26,38 @@
...
-            node_instance_ids=args.get('node_instance_ids'),
-            install_method=args.get('install_methods'))
+    def get(self, _include=None, filters=None, pagination=None, sort=None,
+            all_tenants=None, search=None):
+
+        get_all_results = verify_and_convert_bool(
+            '_get_all_results',
+            request.args.get('_get_all_results', False)
+        )
+        agents_list = get_storage_manager().list(
```
Comment (inbala) [inline rest-service/manager_rest/rest/resources_v3_1/agents.py]: We are adding all this new functionality with filters, sort, include...
We should do one basic test for each one to make sure it works.
You can add the tests to `manager_rest/test/endpoints/test_agents.py`.
Also check the the old api requests works.

File: rest-service/dev-requirements.txt [hunk 10]
```diff
@@ -1,4 +1,4 @@
-git+https://github.com/cloudify-cosmo/cloudify-common@master
+git+https://github.com/cloudify-cosmo/cloudify-common@CY-1447-Improve-agents-list
 
 # For dealing with the binary leftovers of psycopg2 in the 2.7.x version
-psycopg2==2.7.4 --no-binary psycopg2
\ No newline at end of file
+psycopg2==2.7.4 --no-binary psycopg2
```
Comment (inbala) [inline rest-service/dev-requirements.txt]: Make sure to revert the changes
